Separate provenance tracking to different levels

yushangdi opened the pull request:
Summary: as title. We've got request from various parties who are interested in turning on the provenance tracking by default. In this PR, we prepare to turn on part of the provenance tracking that doesn't have too much overhead by default.

- Change `provenance_tracking` config to `provenance_tracking_level` 
- turn on the following provenance tracking by default when `basic_provenance_tracking`=True
    - `set_kernel_post_grad_provenance_tracing` for kernels, this add mapping between triton kernels and post_grad nodes
    - `dump_inductor_provenance_info` if we're dumping tlparse log
    - `get_graph_provenance_json` and dump `reate_mapping_pre_post_grad_nodes`. This creates mapping between pre_grad and post_grad nodes. Since we're not turning on the provenance tracking in GraphTransformObserver by default, the mapping here maybe incomplete/limited. 
    - add stack trace from post grad nodes to inductor IR nodes
    - add exception swallowing for all functions above

Test Plan:
CI

Rollback Plan:

Differential Revision: D80031559




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

yushangdi commented:
> Thanks @yushangdi ! Is there potential latency regression by turning it on by default? can we study them? If so, I can also help study some production models latency implication.

@kflu I don't expect any big latency regression, but it may increase a little.

yushangdi commented on torch/_inductor/compile_fx.py:
```diff
@@ -1055,19 +1058,18 @@ def _compile_fx_inner(
-            )
+    # Dump provenance artifacts for debugging trace
+    if config.trace.basic_provenance_tracking or config.trace.provenance_tracking:
+        trace_structured(
```
we use trace_structured everywhere for tlparse already, it's safe as long as the lambda function passed to it is safe.

yushangdi commented on torch/_inductor/compile_fx.py:
```diff
@@ -1310,20 +1312,13 @@ def codegen_and_compile(
+                    config.trace.basic_provenance_tracking
+                    or config.trace.provenance_tracking
+                ):
                     provenance_tracking_json = (
```
This block should be exception free already. I added exception check in `get_graph_provenance_json` and `create_mapping_pre_post_grad_nodes`

yushangdi commented:
@kflu Would it be possible to also verify whether we can turn on "normal" mode by default in the lowering stack?

kflu commented:
> @kflu Would it be possible to also verify whether we can turn on "normal" mode by default in the lowering stack?

sure, once it's landed we can test both modes.

angelayi commented on torch/_inductor/config.py:
```diff
@@ -1832,10 +1832,18 @@ class trace:
+    # Backward compatibility:
+    #   If TORCH_COMPILE_DEBUG=1, level is set to at least 1.
+    #   If INDUCTOR_PROVENANCE is set, use its integer value.
+    provenance_tracking_level: int = int(
```
I think you can use something like, Literal[0, 1, 2] like [this](https://github.com/pytorch/pytorch/blob/4d38dd14544e4f456df37e361a1022a334c0c52e/torch/_inductor/config.py#L1682-L1684)

yushangdi committed:
Separate provenance tracking to different levels (#160383)

Summary:
as title. We've got request from various parties who are interested in turning on the provenance tracking by default. In this PR, we prepare to turn on part of the provenance tracking that doesn't have too much overhead by default.

- Change `provenance_tracking` config to `provenance_tracking_level`. This is defaults to 1 (normal) now, but will be defaults to 2 to turn on basic provenance tracking
- turn on the following provenance tracking by default when `basic_provenance_tracking`=True
    - `set_kernel_post_grad_provenance_tracing` for kernels, this add mapping between triton kernels and post_grad nodes
    - `dump_inductor_provenance_info` if we're dumping tlparse log
    - `get_graph_provenance_json` and dump `reate_mapping_pre_post_grad_nodes`. This creates mapping between pre_grad and post_grad nodes. Since we're not turning on the provenance tracking in GraphTransformObserver by default, the mapping here maybe incomplete/limited.
    - add exception swallowing for all functions above

Pull Request resolved: https://github.com/pytorch/pytorch/pull/160383

Test Plan:
CI

Rollback Plan: cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

Reviewed By: angelayi

Differential Revision: D80031559

Pulled By: yushangdi
